### DISCUSSION TITLE
Infer generic parameters type from generic constraints

### DISCUSSION THREAD

**Discussion by dczuo** [dotnet/csharplang#5031] [Language Ideas] [answered]
```
public static TControl ItemGenerator<TControl, TItem>(this TControl control, Func<TItem, object?> itemGenerator) where TControl : ItemsControl<TItem>
{
    return control;
}

new ItemsContro<string>().ItemGenerator(item => null);  // currently unable to infer that item is type string
```

**Comment by ZacharyPatten** [comment]
You did not provide enough information. Please elaborate.

Do you mean you have something like this:
```cs
using System;

ItemsControl<string> itemsControl = new();
itemsControl.ItemGenerator<ItemsControl<string>, string>(item => null); // <- line in question

public class ItemsControl<T> { }

public static class Extensions
{
	public static TControl ItemGenerator<TControl, TItem>(this TControl control, Func<TItem, object?> itemGenerator)
		where TControl : ItemsControl<TItem>
	{
		return control;
	}
}
```

And you think you shouldn't have to put `<ItemsControl<string>, string>` like this:
```cs
using System;

ItemsControl<string> itemsControl = new();
itemsControl.ItemGenerator(item => null); // <- line in question

public class ItemsControl<T> { }

public static class Extensions
{
	public static TControl ItemGenerator<TControl, TItem>(this TControl control, Func<TItem, object?> itemGenerator)
		where TControl : ItemsControl<TItem>
	{
		return control;
	}
}
```

**Reply by ZacharyPatten** [reply]
Depending on the actual topic of this discussion, it may be related to: https://github.com/dotnet/csharplang/issues/3817

**Reply by dczuo** [reply]
Thank you for your reply and for providing a more complete code

```cs
using System;

ItemsControl<string> itemsControl = new();
itemsControl.ItemGenerator<ItemsControl<string>, string>(item => null); // Ok
itemsControl.ItemGenerator((string item) => null); // Ok
itemsControl.ItemGenerator(item => null);  // CS0411. My question is whether it is possible to infer from "where TControl : ItemsControl<TItem>" that TItem is type string

public class ItemsControl<T> { }

public static class Extensions
{
	public static TControl ItemGenerator<TControl, TItem>(this TControl control, Func<TItem, object?> itemGenerator)
		where TControl : ItemsControl<TItem>
	{
		return control;
	}
}
```

**Reply by ZacharyPatten** [reply]
This issue is predicated on the idea that there is only one possible type that `<T2>` (or in your example `<TItem>`) could be, and that predicate is not always true. Here is an example where it is not true:
```cs
using System;

A a = new();
B b = new();

Console.WriteLine(a is I<string>);
Method<A, string>(a);
//Method(a); // <- currently "T2" can only be "string" for "A" types,
             //    but if the class hierarchy was changed, this it could
             //    allow other types such as the "B" example below

Console.WriteLine(b is I<string>);
Console.WriteLine(b is I<int>);
Method<B, string>(b);
Method<B, int>(b);

static void Method<T1, T2>(T1 t1)
	where T1 : I<T2>
{
	Console.WriteLine(typeof(T1).Name);
	Console.WriteLine(typeof(T2).Name);
}

public interface I<T> { }
public class A : I<string> { }
public class B : A, I<int> { }
```

The `B` type implements both `I<string>` and `I<int>` and thus when used as `<T1>` in `Method`, it means that `<T2>` could be `string` or `int` and thus cannot be infered by the compiler.

In this example, when `A` is used as `<T1>` in `Method`, `<T2>` can only be `string` since `A` only implements `I<string>`, but having the compiler infer that would be a bad idea, because it means that you can break that method call if you change `A` so that it implements other types. Having scenarios where you can break a method by adding an interface implementation is very undesireable from a language design perspective in my opinion.

**Comment by HaloFour** [comment, marked as answer]
I want to say that there are a couple of existing discussions on this subject.  This looks like one, which references a few others and also has some comments from the team regarding an exploration of how it could be implemented: https://github.com/dotnet/csharplang/discussions/997
